### PR TITLE
Optimize ReplicationController and ControllerRef

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/30759

@lavalamp @gmarek

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30768)

### CODE REVIEW

**Comment by caesarxuchao** [comment]
Is this outdated by https://github.com/kubernetes/kubernetes/pull/30774?

**Comment by wojtek-t** [comment]
It's not outdated - but we probably should merge the other one first.

**Comment by wojtek-t** [comment]
I would also like to do the same for ReplicaSet.

**Comment by caesarxuchao** [comment]
yeah, I missed the meat, I'm taking a look now.

File: pkg/controller/replication/replication_controller.go [hunk 1]
```diff
@@ -237,6 +237,38 @@ func (rm *ReplicationManager) Run(workers int, stopCh <-chan struct{}) {
 // TODO: Surface that we are ignoring multiple controllers for a single pod.
 // TODO: use ownerReference.Controller to determine if the rc controls the pod.
 func (rm *ReplicationManager) getPodController(pod *api.Pod) *api.ReplicationController {
+	if rm.garbageCollectorEnabled {
+		controllerRef := controller.GetControllerOf(&pod.ObjectMeta)
```
Comment (caesarxuchao) [inline pkg/controller/replication/replication_controller.go]: Thanks for doing this. I guess the improvement is not as big as 30774, given we already had a rc lookup cache. So let's merge 30774 first.

File: pkg/controller/replication/replication_controller.go [hunk 2]
```diff
@@ -237,6 +237,38 @@ func (rm *ReplicationManager) Run(workers int, stopCh <-chan struct{}) {
 // TODO: Surface that we are ignoring multiple controllers for a single pod.
 // TODO: use ownerReference.Controller to determine if the rc controls the pod.
```
Comment (kargakis) [inline pkg/controller/replication/replication_controller.go]: remove this todo?

**Comment by wojtek-t** [comment]
@caesarxuchao - so the issue with the failing integration test is real. Basically, if pod label changes (and it isn't owner by any controller), we shouldn't use controllerRef to determine the controller, because it will never be adopted.
However, there are probably other cases when this can happen (e.g. when other controller is orphaning a pod, the its controllerRef will be empty, and thus we also shouldn't use it for adoption).

That said, I'm not convinced that we can do anything reasonable with this PR.
All I can think of for now is that, we may just try to use it as optimization - try to check "ControllerRef" - if that's non-empty, just use it. Otherwise fallback to operating on selectors.
WDYT?

**Comment by caesarxuchao** [comment]
> However, there are probably other cases when this can happen (e.g. when other controller is orphaning a pod, the its controllerRef will be empty, and thus we also shouldn't use it for adoption).

Right. If possible, could you add a integration test for this case?

>  try to check "ControllerRef" - if that's non-empty, just use it. Otherwise fallback to operating on selectors.
 
Sounds right.

**Pull request closed without merging**